// dex-parse/index.js
"use strict";

var fs = require('fs')

class DexData {
    constructor() {
        this.products = [];
        this.events = [];
    }

    startEvent(name) {
      this.currEvent = {}
      this.events.push(this.currEvent)
    }

    startItem(name) {
        this.currItem = {}
        this.products.push(this.currItem)
    }

    ensureEvent() {
      if (this.currEvent == undefined) {
        this.startEvent
      }
    }

    ensureItem() {
        if(this.currItem == undefined) {
            this.startItem()
        }
    }

    complete() {
        delete this.currEvent
        delete this.currItem
    }
}

function assign(parts, fieldNumber, object, property, lambda) {
    var fieldValue = parts[fieldNumber]
    if(typeof(fieldValue) == 'string') {
        fieldValue = fieldValue.trim()
        if(lambda != undefined) {
            object[property] = lambda(fieldValue)
        } else {
            object[property] = fieldValue
        }
    }
}

exports.DexData = DexData;

var parseID1 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}

    assign(parts, 1, context.machine, "serialNumber")
    assign(parts, 2, context.machine, "modelNumber")
    assign(parts, 3, context.machine, "buildStandard")
    assign(parts, 6, context.machine, "assetNumber")
}

var parseID4 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}

    assign(parts, 1, context.machine, "decimalPointPosition")
    assign(parts, 2, context.machine, "numericCurrencyCode")
    assign(parts, 3, context.machine, "alphabeticCurrencyCode")
}

var parseID5 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}

    assign(parts, 1, context.machine, "systemDate")
    assign(parts, 2, context.machine, "systemTime")
    assign(parts, 3, context.machine, "systemTimeSeconds")
    assign(parts, 4, context.machine, "systemDaylightSavingMode")
}

var parseCB1 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.controlBoard = context.machine.controlBoard || {}

    assign(parts, 1, context.machine.controlBoard, "serialNumber")
    assign(parts, 2, context.machine.controlBoard, "modelNumber")
    assign(parts, 3, context.machine.controlBoard, "softwareRevision")
}

var parseVA1 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.valueAdded = context.machine.valueAdded || {}

    assign(parts, 1, context.machine.valueAdded, "valueOfAllPaidVendsSinceInstallation")
    assign(parts, 2, context.machine.valueAdded, "numberOfAllPaidVendsSinceInstallation")
    assign(parts, 3, context.machine.valueAdded, "valueOfAllPaidVendsSinceLastReset")
    assign(parts, 4, context.machine.valueAdded, "numberOfAllPaidVendsSinceLastReset")
    assign(parts, 5, context.machine.valueAdded, "valueOfAllDiscountGivenSinceInstallation")
    assign(parts, 6, context.machine.valueAdded, "numberOfAllDiscountGivenSinceInstallation")
    assign(parts, 7, context.machine.valueAdded, "valueOfAllDiscountGivenSinceLastReset")
    assign(parts, 8, context.machine.valueAdded, "numberOfAllDiscountGivenSinceLastReset")
    assign(parts, 9, context.machine.valueAdded, "numberofAllSurchargedVendsSinceInstallation")
    assign(parts, 10, context.machine.valueAdded, "valueofAllSurchargedVendsSinceInstallation")
    assign(parts, 11, context.machine.valueAdded, "numberofAllSurchargedVendsSinceLastReset")
    assign(parts, 12, context.machine.valueAdded, "valueofAllSurchargedVendsSinceLastReset")
}

var parseVA2 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.valueAdded = context.machine.valueAdded || {}

    assign(parts, 1, context.machine.valueAdded, "testVendValueOfSalseSinceInstallation")
    assign(parts, 2, context.machine.valueAdded, "numberOfTestVendSinceInstallation")
    assign(parts, 3, context.machine.valueAdded, "testVendValueOfSalseSinceLastReset")
    assign(parts, 4, context.machine.valueAdded, "numberOfTestVendSinceLastReset")
    assign(parts, 5, context.machine.valueAdded, "testVendBoxCashSinceInstallation")
    assign(parts, 6, context.machine.valueAdded, "testVendBoxCashSinceLastReset")
}

var parseVA3 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.valueAdded = context.machine.valueAdded || {}

    assign(parts, 1, context.machine.valueAdded, "freeVendValueOfSalesSinceInstallation")
    assign(parts, 2, context.machine.valueAdded, "numberOfFreeVendSinceInstallation")
    assign(parts, 3, context.machine.valueAdded, "freeVendValueOfSalesSinceLastRest")
    assign(parts, 4, context.machine.valueAdded, "numberOfFreeVendSinceLastReset")
}

var parseCA1 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "coinMechanismSerialNumber")
    assign(parts, 2, context.machine.cashAttribute, "coinMechanismModelNumber")
    assign(parts, 3, context.machine.cashAttribute, "coinMechanismSoftwareRevision")
    assign(parts, 4, context.machine.cashAttribute, "userDefinedFieldCA1")
    assign(parts, 5, context.machine.cashAttribute, "coinMechanismAssetNumber")
}

var parseCA2 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashSalesSinceInstallation")
    assign(parts, 2, context.machine.cashAttribute, "numberOfCashVendsSinceInstallation")
    assign(parts, 3, context.machine.cashAttribute, "valueOfCashSalesSinceLastReset")
    assign(parts, 4, context.machine.cashAttribute, "numberOfCashVendsSinceLastReset")
}

var parseCA3 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashInSinceLastReset")
    assign(parts, 2, context.machine.cashAttribute, "valueOfCashToCashBoxSinceLastReset")
    assign(parts, 3, context.machine.cashAttribute, "valueOfCashToTubesSinceLastReset")
    assign(parts, 4, context.machine.cashAttribute, "valueOfBillsInSinceLastReset")
    assign(parts, 5, context.machine.cashAttribute, "valueOfCashInSinceInstallation")
    assign(parts, 6, context.machine.cashAttribute, "valueOfCashToCashBoxSinceInstallation")
    assign(parts, 7, context.machine.cashAttribute, "valueOfCashToTubesSinceInstallation")
    assign(parts, 8, context.machine.cashAttribute, "valueOfBillsInSinceInstallation")
    assign(parts, 9, context.machine.cashAttribute, "valueOfBillsInSinceLastReset")
    assign(parts, 10, context.machine.cashAttribute, "valueOfBillsInSinceInstallation")
    assign(parts, 11, context.machine.cashAttribute, "valueOfBillsToRecyclerSinceLastReset")
    assign(parts, 12, context.machine.cashAttribute, "valueOfBillsToRecyclerSinceInstallation")
}

var parseCA4 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashDispensedSinceLastReset")
    assign(parts, 2, context.machine.cashAttribute, "valueOfCashManuallyDispensedSinceLastReset")
    assign(parts, 3, context.machine.cashAttribute, "valueOfCashDispensedSinceInstallation")
    assign(parts, 4, context.machine.cashAttribute, "valueOfCashManuallyDispensedSinceInstallation")
    assign(parts, 5, context.machine.cashAttribute, "valueOfBillsDispensedSinceLastReset")
    assign(parts, 6, context.machine.cashAttribute, "valueOfBillsManuallyDispensedSinceLastReset")
    assign(parts, 7, context.machine.cashAttribute, "valueOfBillsDispensedToRecyclerSinceLastReset")
    assign(parts, 8, context.machine.cashAttribute, "valueOfBillsDispensedSinceInstallation")
    assign(parts, 9, context.machine.cashAttribute, "valueOfBillsManuallyDispensedSinceInstallation")
    assign(parts, 10, context.machine.cashAttribute, "valueOfBillsManuallyDispensedToRecyclerSinceInstallation")
}

var parseCA7 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashDiscountsSinceLastReset")
    assign(parts, 2, context.machine.cashAttribute, "valueOfCashDiscountsSinceInstallation")
    assign(parts, 3, context.machine.cashAttribute, "numberOfCashDiscountsSinceLastReset")
    assign(parts, 4, context.machine.cashAttribute, "numberOfCashDiscountsSinceInstallation")
    assign(parts, 5, context.machine.cashAttribute, "valueOfCashSurchargesSinceLastReset")
    assign(parts, 6, context.machine.cashAttribute, "valueOfCashSurchargesSinceInstallation")
    assign(parts, 7, context.machine.cashAttribute, "numberOfCashSurchargesSinceLastReset")
    assign(parts, 8, context.machine.cashAttribute, "numberOfCashSurchargesSinceInstallation")
}

var parseCA8 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashOverpaySinceInstallation")
    assign(parts, 2, context.machine.cashAttribute, "valueOfCashOverpaySinceLastReset")
}

var parseCA10 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueOfCashFilledSinceLastReset")
    assign(parts, 2, context.machine.cashAttribute, "valueOfCashFilledSinceInstallation")
    assign(parts, 3, context.machine.cashAttribute, "valueOfBillsFilledSinceLastReset")
    assign(parts, 4, context.machine.cashAttribute, "valueOfBillsFilledSinceInstallation")
}

var parseCA15 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashAttribute = context.machine.cashAttribute || {}

    assign(parts, 1, context.machine.cashAttribute, "valueofTubeContents")
    assign(parts, 2, context.machine.cashAttribute, "blockNumber")
    assign(parts, 3, context.machine.cashAttribute, "coinType0Or7")
    assign(parts, 4, context.machine.cashAttribute, "coinType1Or8")
    assign(parts, 5, context.machine.cashAttribute, "coinType2Or9")
    assign(parts, 6, context.machine.cashAttribute, "coinType3Or10")
    assign(parts, 7, context.machine.cashAttribute, "coinType4Or11")
    assign(parts, 8, context.machine.cashAttribute, "coinType5Or12")
    assign(parts, 9, context.machine.cashAttribute, "coinType6Or13")
    assign(parts, 10, context.machine.cashAttribute, "coinType7Or14")
}

var parseBA1 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.billValidator = context.machine.billValidator || {}

    assign(parts, 1, context.machine.billValidator, "billValidatorSerialNumber")
    assign(parts, 2, context.machine.billValidator, "billValidatorModelNumber")
    assign(parts, 3, context.machine.billValidator, "billValidatorSoftwareRevision")
    assign(parts, 4, context.machine.billValidator, "userDefinedFieldBA1")
    assign(parts, 5, context.machine.billValidator, "billValidatorAssetNumber")
}

var parseDA1 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless = context.machine.cashless || {}

    assign(parts, 1, context.machine.cashless, "cashless1SerialNumber")
    assign(parts, 2, context.machine.cashless, "cashless1ModelNumber")
    assign(parts, 3, context.machine.cashless, "cashless1SoftwareRevision")
    assign(parts, 4, context.machine.cashless, "userDefinedFieldDA1")
    assign(parts, 5, context.machine.cashless, "cashless1AssetNumber")
}

var parseDA2 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless = context.machine.cashless || {}

    assign(parts, 1, context.machine.cashless, "valueOfCashless1SalesSinceInstallation")
    assign(parts, 2, context.machine.cashless, "numberOfCashless1VendsSinceInstallation")
    assign(parts, 3, context.machine.cashless, "valueOfCashless1SalesSinceLastReset")
    assign(parts, 4, context.machine.cashless, "numberOfCashless1VendsSinceLastReset")
    assign(parts, 5, context.machine.cashless, "userDefinedFieldDA2")
}

var parseDA4 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless = context.machine.cashless || {}

    assign(parts, 1, context.machine.cashless, "valueCreditedToCashless1SinceInstallation")
    assign(parts, 2, context.machine.cashless, "valueCreditedToCashless1SinceLastReset")
    assign(parts, 3, context.machine.cashless, "userDefinedFieldDA4")
}

var parseDA5 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless = context.machine.cashless || {}

    assign(parts, 1, context.machine.cashless, "valueOfCashless1DiscountsSinceLastReset")
    assign(parts, 2, context.machine.cashless, "numberOfDiscountCashless1VendsSinceLastReset")
    assign(parts, 3, context.machine.cashless, "valueOfCashless1DiscountsSinceInstallation")
    assign(parts, 4, context.machine.cashless, "numberOfDiscountCashless1VendsSinceInstallation")
    assign(parts, 5, context.machine.cashless, "valueOfCashless1SurchargesSinceLastReset")
    assign(parts, 6, context.machine.cashless, "numberOfSurchargeCashless1VendsSinceLastReset")
    assign(parts, 7, context.machine.cashless, "valueOfCashless1SurchargesSinceInstallation")
    assign(parts, 8, context.machine.cashless, "numberOfSurchargeCashless1VendsSinceInstallation")
}

var parseDA7 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless = context.machine.cashless || {}

    assign(parts, 1, context.machine.cashless, "cashless1UserGroupNumber")
    assign(parts, 2, context.machine.cashless, "cashless1UserGroupNetSalesSinceLastReset")
    assign(parts, 3, context.machine.cashless, "userGroupValueAddedToCashless1SinceLastReset")
    assign(parts, 4, context.machine.cashless, "cashless1UserGroupDiscountsSinceLastReset")
    assign(parts, 5, context.machine.cashless, "numberOfProductsSoldToCashless1UserGroupSinceInstallation")
    assign(parts, 6, context.machine.cashless, "numberOfProductsSoldToCashless1UserGroupSinceLastReset")
    assign(parts, 7, context.machine.cashless, "cashless1UserGroupNetSalesSinceInstallation")
    assign(parts, 8, context.machine.cashless, "userGroupValueAddedToCashless1SinceInstallation")
    assign(parts, 9, context.machine.cashless, "cashless1UserGroupDiscountsSinceInstallation")
}

var parseDB1 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless2 = context.machine.cashless2 || {}

    assign(parts, 1, context.machine.cashless2, "cashless2SerialNumber")
    assign(parts, 2, context.machine.cashless2, "cashless2ModelNumber")
    assign(parts, 3, context.machine.cashless2, "cashless2SoftwareRevision")
    assign(parts, 4, context.machine.cashless2, "userDefinedFieldDB1")
    assign(parts, 5, context.machine.cashless2, "cashless2AssetNumber")
}

var parseDB2 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless2 = context.machine.cashless2 || {}

    assign(parts, 1, context.machine.cashless2, "valueOfCashless2SalesSinceInstallation")
    assign(parts, 2, context.machine.cashless2, "numberOfCashless2VendsSinceInstallation")
    assign(parts, 3, context.machine.cashless2, "valueOfCashless2SalesSinceLastReset")
    assign(parts, 4, context.machine.cashless2, "numberOfCashless2VendsSinceLastReset")
    assign(parts, 5, context.machine.cashless2, "userDefinedFieldDB2")
}

var parseDB4 = function(line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless2 = context.machine.cashless2 || {}

    assign(parts, 1, context.machine.cashless2, "valueCreditedToCashless2SinceInstallation")
    assign(parts, 1, context.machine.cashless2, "valueCreditedToCashless2SinceLastReset")
    assign(parts, 1, context.machine.cashless2, "userDefinedFieldDB4")
}

var parseDB5 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.cashless2 = context.machine.cashless2 || {}

    assign(parts, 1, context.machine.cashless2, "valueOfCashless2DiscountsSinceLastReset")
    assign(parts, 2, context.machine.cashless2, "numberOfDiscountCashless2VendsSinceLastReset")
    assign(parts, 3, context.machine.cashless2, "valueOfCashless2DiscountsSinceInstallation")
    assign(parts, 4, context.machine.cashless2, "numberOfDiscountCashless2VendsSinceInstallation")
    assign(parts, 5, context.machine.cashless2, "valueOfCashless2SurchargesSinceLastReset")
    assign(parts, 6, context.machine.cashless2, "numberOfSurchargeCashless2VendsSinceLastReset")
    assign(parts, 7, context.machine.cashless2, "valueOfCashless2SurchargesSinceInstallation")
    assign(parts, 8, context.machine.cashless2, "numberOfSurchargeCashless2VendsSinceInstallation")
}

var parseTA1 = function (line, context) {
    var parts = line.split("*")
    context.machine = context.machine || {}
    context.machine.tokenAcceptor = context.machine.tokenAcceptor || {}

    assign(parts, 1, context.machine.tokenAcceptor, "tokenAcceptorSerialNumber")
    assign(parts, 2, context.machine.tokenAcceptor, "tokenAcceptorModelNumber")
    assign(parts, 3, context.machine.tokenAcceptor, "tokenAcceptorSoftwareRevision")
    assign(parts, 4, context.machine.tokenAcceptor, "userDefinedFieldTA1")
}

var parseEA1 = function (line, context) {
    context.startEvent()
    var parts = line.split("*")

    assign(parts, 1, context.currEvent, "eventIdentification")
    assign(parts, 2, context.currEvent, "dataOfEventOccurance")
    assign(parts, 3, context.currEvent, "timeOfEventOccurance")
    assign(parts, 4, context.currEvent, "eventDurationInMinutes")
    assign(parts, 5, context.currEvent, "eventDurationInMilliseconds")
}

var parseEA2 = function (line, context) {
    context.ensureEvent()
    var parts = line.split("*")

    assign(parts, 1, context.currEvent, "eventIdentification")
    assign(parts, 2, context.currEvent, "numberOfEventsSinceLastReset")
    assign(parts, 3, context.currEvent, "numberOfEventsSinceInstallation")
    assign(parts, 4, context.currEvent, "userDefinedFieldEA2")
    assign(parts, 5, context.currEvent, "eventActivity")
    assign(parts, 6, context.currEvent, "eventDurationInMinutes")
}


var parsePA1 = function(line, context) {
    context.startItem()
    var parts = line.split("*")
    assign(parts, 1, context.currItem, "name")
    assign(parts, 2, context.currItem, "price", (i) => i / 100)

}

var parsePA2 = function(line, context) {
    context.ensureItem()
    var parts = line.split("*")
    assign(parts, 1, context.currItem, "sold", (i) => Number(i))
    assign(parts, 2, context.currItem, "revenue", (i) => i / 100)
}


var parsePA3 = function(line, context) {
    context.ensureItem()
    var parts = line.split("*")
    assign(parts, 1, context.currItem, "testVendCount")
}

var parsePA5 = function(line, context) {
    context.ensureItem()
    var parts = line.split("*")
    assign(parts, 1, context.currItem, "soldOutDate")
    assign(parts, 2, context.currItem, "soldOutTime")
    assign(parts, 3, context.currItem, "soldOutCount")
}



var defaultHandlers = {
    "ID1" : parseID1,
    "ID4" : parseID4,
    "ID5" : parseID5,
    "CB1" : parseCB1,
    "VA1" : parseVA1,
    "VA2" : parseVA2,
    "VA3" : parseVA3,
    "CA1" : parseCA1,
    "CA2" : parseCA2,
    "CA7" : parseCA7,
    "CA8" : parseCA8,
    "CA10": parseCA10,
    "CA15": parseCA15,
    "BA1" : parseBA1,
    "DA1" : parseDA1,
    "DA2" : parseDA2,
    "DA4" : parseDA4,
    "DA5" : parseDA5,
    "DA7" : parseDA7,
    "DB1" : parseDB1,
    "DB2" : parseDB2,
    "DB4" : parseDB4,
    "DB5" : parseDB5,
    "EA1" : parseEA1,
    "EA2" : parseEA2,
    "PA1" : parsePA1,
    "PA2" : parsePA2,
    "PA3" : parsePA3,
    "PA5" : parsePA5
};

exports.readText = function(text, cb) {
    var handlers = defaultHandlers

    text = text.toString()

    if(text.length <= 0) {
        cb(new Error('file or text was empty'));
        return
    }

    var lines = text.toString().split('\n')

    var dexdata = new DexData()

    lines.forEach(function(line) {
        line = line.replace("\r", "")
        var prefix2 = line.substring(0, 2)
        var prefix3 = line.substring(0, 3)
        var prefix4 = line.substring(0, 4)

        var handler = handlers[prefix4] || handlers[prefix3] || handlers[prefix2]

        if(handler) {
            handler(line, dexdata)
        }
    })

    dexdata.complete()

    cb(undefined, dexdata);
};

exports.readFile = function(path, cb) {
    fs.readFile(path, (err, data) => {
        if(err) {
            cb(err);
        };

        exports.readText(data, cb);
    })
};
